controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+
+		cluster, err := util.GetClusterFromMetadata(ctx, r.Client, claim.ObjectMeta)
+		if err != nil {
+			log.Info("Could not get cluster resource for IPAddressClaim")
+			return ctrl.Result{}, nil
+		}
+
+		infraCluster, err := r.getInfraCluster(ctx, cluster, &claim)
+		if err != nil {
+			return ctrl.Result{}, errors.New("error getting infra provider cluster")
+		} else if infraCluster == nil {
+			log.Info("infra cluster is not ready yet")
+			return ctrl.Result{}, nil
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			clusterName := fmt.Sprintf("%s-%s", claim.ObjectMeta.Labels[clusterv1.ClusterNameLabel], claim.Namespace)
+			ip, err := r.getIP(ctx, scope, pool, infraCluster, clusterName)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress := &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+				},
+			}
+
+			if !contains(pool.Spec.PreAllocatedFloatingIPs, ip) && pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
+			if err = r.Client.Status().Update(ctx, &claim); err != nil {
+				log.Error(err, "Failed to update IPAddressClaim status", "claim", claim.Name, "ip", ip)
+				return ctrl.Result{}, err
+			}
+			scope.Logger().Info("Claimed IP", "ip", ip)
+		}
+	}
+	if err = r.setIPStatuses(ctx, scope, pool); err != nil {
```

Comment:
I'm not sure I'd bother with this. Any operation which altered ip statuses would result in us being reconciled again immediately, in which case we'd update this on the next reconcile anyway.